cfe_internal/enterprise/mission_portal.cf

```diff
@@ -228,7 +228,7 @@ bundle agent mission_portal_apache_from_stage(config, staged_config)
                            "place only if we have stopped apache already.");
 
       "LD_LIBRARY_PATH=$(sys.workdir)/lib:$LD_LIBRARY_PATH $(sys.workdir)/httpd/bin/apachectl" -> { "ENT-9686" }
-        args => "restart",
+        args => "graceful",
```

Comment:
Is this synchronous or async? I was thinking graceful is async. If it's async then whatever changes we expected to take effect from the apache reload won't be in effect at the end of this promise execution. I dunno if that is a problem, but it was something that came to mind.

Reply:
Good point. I will check into it and/or test it.

Reply:
from [URL](https://httpd.apache.org/docs/current/programs/apachectl.html)

```
The apachectl script returns a 0 exit value on success, and >0 if an error occurs. For more details, view the comments in the script.
```

It would seem it is sync.

Also interestingly it checks the config so it would seem we don't need to do that in policy as well?

```
This command automatically checks the configuration files as in configtest before initiating the restart to make sure Apache doesn't die.
```

This checking of the configuration files as in configtest is the same behavior for both restart and graceful according to the docs mentioned above.

I will test with bad config and see what happens.

Reply:
We still need to know that the config is broken if that is the case, right?

Reply:
> I dunno if that is a problem

I don't think it should be a problem. Even if it's async, I think it guarantees that **new** connections will be handled according to the new config when they are actually processed.

Reply:
@nickanderson @vpodzime I tried our patched /var/cfengine/httpd/bin/apachectl out and it seems graceful is as synchronous as the restart command. It will fail if config is broken like this:

```
root@ubuntu2004:/var/cfengine/share/GUI# /var/cfengine/httpd/bin/apachectl graceful
AH00526: Syntax error on line 1 of /var/cfengine/httpd/conf/httpd.conf:
Invalid command 'BROKEN', perhaps misspelled or defined by a module not included in the server configuration
root@ubuntu2004:/var/cfengine/share/GUI# echo $?
1
```

Looking through the source code at https://github.com/apache/httpd it seems that apachectl/httpd -k merely send good ol' UNIX signals to the httpd process.

in server/mpm_unix.c
```

    if (!strcmp(dash_k_arg, "graceful")) {
        if (!running) {
            printf("httpd not running, trying to start\n");
        }
        else {
            *exit_status = send_signal(otherpid, AP_SIG_GRACEFUL);
            return 1;
        }
    }
```

and send_signal() just returns the result of a kill() call

```
static int send_signal(pid_t pid, int sig)
{
    if (kill(pid, sig) < 0) {
        ap_log_error(APLOG_MARK, APLOG_STARTUP, errno, NULL, APLOGNO(00057)
                     "sending signal to server");
        return 1;
    }
    return 0;
}
```

So we aren't getting back any information about httpd or it's state other than if kill() succeeds then we know that apachectl/httpd could find the proper PID and was able to send the signal.